backend/app/synthesize_polygons/synthesize_polygon.py

```diff
@@ -103,29 +123,56 @@ def generate_note_with_amplitude(frequency, duration, amplitude):
     :return: numpy array which represents the note
     """
     time_steps = np.linspace(0, duration, int(duration * WAV_SAMPLE_RATE), False)
-    note = np.sin(frequency * time_steps * 2 * np.pi) * amplitude
+    note = np.sin(frequency * time_steps * 2 * np.pi) * amplitude * amplitude_decay(frequency)
     return note
 
 
+def generate_time_stamps(duration_list, note_delay, sides_as_duration=False):
+    """
+    Generates a list indicating when the sound associated with each line of the polygon should
+    start and end in seconds.
+    :param duration_list: a list of durations of each side
+    :param note_delay: delay between each note in seconds
+    :param sides_as_duration: whether to use side lengths to determine duration. if False, then use
+    side lengths to determine amplitude.
+    :return: list of tuples signifying the time in which each line starts and ends in seconds
+    """
+
+    time_stamps = []
+    start = 0
+    for duration in duration_list:
+        end = start + duration
+        time_stamps.append((start, end))
+        if sides_as_duration:
+            start += duration
+        else:
+            start += note_delay
+
+    return time_stamps
+
+
 # pylint: disable=too-many-arguments
 # pylint: disable=too-many-locals
-# TODO: Fill in for and use restrict_octave
-# TODO: Fill in for and use sides_as_duration along with sides_as_duration function
-def synthesize_polygon(points, note_length=1, note_delay=0, restrict_octave=False,
-                       sides_as_duration=False, base_frequency=220):
+def synthesize_polygon(points, note_length=1, note_delay=1, restrict_frequency=False,
+                       sides_as_duration=False, base_frequency=220, floor_frequency=20,
+                       ceil_frequency=10000):
     """
     Synthesizes a polygon. The polygon is represented as a list of points
     where each point is a tuple of length 2.
     :param points: list of points representing a polygon.
     :param note_length: length of each note in seconds.
     :param note_delay: delay between each note in seconds.
-    :param restrict_octave: whether to restrict the notes to a single octave
+    :param restrict_frequency: whether to restrict the notes to a specified frequency range
     :param sides_as_duration: whether to use side lengths to determine duration. if False, then use
     side lengths to determine amplitude.
     :param base_frequency: the frequency of the first note of the polygon
+    :param floor_frequency: the lowest allowable frequency for any note
+    :param ceil_frequency: the highest allowable frequency for any note
     :return: numpy array which represents the sound.
     """
     check_valid_polygon(points)
+    assert ceil_frequency >= 2*floor_frequency, \
```

Comment:
Fine for now, but as a TODO for later cleanup work: instead of bare asserts in the backend, if these are states that are possible to reach through the UI, we want to do some error handling here and report the error to the frontend.

A straightforward technique would be to replace this with something like this:

```
if ceil_frequency >= 2 * floor_frequency:
    raise ValueError(ERROR_MSG)
```

... and then in the calling view function that returns a response, you can either do something like throw an [APIException](https://www.django-rest-framework.org/api-guide/exceptions/) or write your own error reporting by returning a response with an [appropriate error code](https://www.django-rest-framework.org/api-guide/status-codes/).

And then you need to respond to the error in the frontend!

So: a non-trivial project to do this error handling right, and we're not going to mess with it for right now, but just FYI.